database/src/backend/heed/database.rs

```diff
@@ -26,72 +26,127 @@ use crate::{
 /// An opened read-only database associated with a transaction.
 ///
 /// Matches `redb::ReadOnlyTable`.
-pub(super) struct HeedTableRo<'env, T: Table> {
+pub(super) struct HeedTableRo<'tx, T: Table> {
     /// An already opened database table.
-    db: HeedDb<T::Key, T::Value>,
+    pub(super) db: HeedDb<T::Key, T::Value>,
     /// The associated read-only transaction that opened this table.
-    tx_ro: &'env heed::RoTxn<'env>,
+    pub(super) tx_ro: &'tx heed::RoTxn<'tx>,
 }
 
 /// An opened read/write database associated with a transaction.
 ///
 /// Matches `redb::Table` (read & write).
-pub(super) struct HeedTableRw<'env, T: Table> {
+pub(super) struct HeedTableRw<'env, 'tx, T: Table> {
     /// TODO
-    db: HeedDb<T::Key, T::Value>,
+    pub(super) db: HeedDb<T::Key, T::Value>,
     /// The associated read/write transaction that opened this table.
-    tx_rw: &'env mut heed::RwTxn<'env>,
+    pub(super) tx_rw: &'tx mut heed::RwTxn<'env>,
+}
+
+//---------------------------------------------------------------------------------------------------- Shared functions
+// FIXME: we cannot just deref `HeedTableRw -> HeedTableRo` and
+// call the functions since the database is held by value, so
+// just use these generic functions that both can call instead.
+
+/// Shared generic `get()` between `HeedTableR{o,w}`.
+#[inline]
+fn get<'tx, T: Table>(
+    db: &'_ HeedDb<T::Key, T::Value>,
+    tx_ro: &'tx heed::RoTxn<'_>,
+    key: &T::Key,
+) -> Result<impl Borrow<T::Value> + 'tx, RuntimeError> {
+    match db.get(tx_ro, key) {
+        Ok(Some(value)) => Ok(value),
+        Ok(None) => Err(RuntimeError::KeyNotFound),
+        Err(e) => Err(e.into()),
+    }
+}
+
+/// Shared generic `get_range()` between `HeedTableR{o,w}`.
+#[inline]
+#[allow(clippy::needless_pass_by_value, clippy::trait_duplication_in_bounds)]
+fn get_range<'tx, T: Table, Range>(
+    db: &'tx HeedDb<T::Key, T::Value>,
+    tx_ro: &'tx heed::RoTxn<'_>,
+    range: Range,
+) -> Result<impl Iterator<Item = Result<impl Borrow<T::Value> + 'tx, RuntimeError>>, RuntimeError>
+where
+    Range: RangeBounds<T::Key> + RangeBounds<&'tx T::Key> + 'tx,
+{
+    /// TODO
+    struct Iter<'tx, T: Table> {
+        /// TODO
+        iter: heed::RoRange<'tx, StorableHeed<T::Key>, StorableHeed<T::Value>>,
+    }
+
+    // TODO
+    impl<'tx, T: Table> Iterator for Iter<'tx, T> {
+        type Item = Result<&'tx T::Value, RuntimeError>;
+        fn next(&mut self) -> Option<Self::Item> {
+            // TODO
+            self.iter
+                .next()
+                .map(|result| result.map(|value| value.1).map_err(RuntimeError::from))
+        }
+    }
+
+    // TODO
+    match db.range(tx_ro, &range) {
+        Ok(iter) => Ok(Iter::<'_, T> { iter }),
+        Err(e) => Err(RuntimeError::from(e)),
+    }
```

Comment:
I don't think you need a custom iterator here:

```suggestion
    Ok(db.range(tx_ro, &range)?.map(|res| Ok(res?.1)))
```